In [1]:
import pandas as pd
import json

# Validation on dataset 2

In [2]:
OUTPUT_DIR = "../003-Evaluating-models/output_csvs"
        
df1 = pd.read_csv(f"{OUTPUT_DIR}/eval_trained_on_dataset_v2_task_val_ds_original.csv")

original_cols = ['model_key', 'model', 'transfer_learning', 'imgsz', 'epochs', 'batch',
               'dataset', 'optimizer', 'lr0', 'all_P', 'all_R', 'all_F1', 'all_mAP@.5',
               'all_mAP@.5:.95']

new_cols = ['model_key', 'model', 'TL', 'imgsz', 'epochs', 'batch',
           'dataset', 'optimizer', 'lr0', 'Precision', 'Recall', 'all_F1', 'mAP@.5',
           'mAP@.5:.95']

df1.rename(columns=dict(zip(original_cols, new_cols)), inplace=True)
df1 = df1.loc[[2,3,1,7,9]]

In [3]:
# Load data from the JSON file
with open('optimization_results_on_v2.json', 'r') as json_file:
    data = json.load(json_file)
    
# Create a Pandas DataFrame
df2 = pd.DataFrame(data)

original_cols = ['Model', 'TL', 'imgsz', 'epochs', 'batch', 'dataset', 'optimizer', 'lr',
                   'optimization_technique', 'model_performance', 'Precision', 'Recall',
                   'mAP@.5', 'mAP@.5:.95', 'Class', 'Images', 'Labels']

new_cols = ['Model', 'TL', 'imgsz', 'epochs', 'batch', 'dataset', 'optimizer', 'lr',
                   'optimization_technique', 'FPS', 'Precision', 'Recall',
                   'mAP@.5', 'mAP@.5:.95', 'Class', 'Images', 'Labels']

df2.rename(columns=dict(zip(original_cols, new_cols)), inplace=True)

In [4]:
df_rc = pd.read_csv(f"{OUTPUT_DIR}/eval_trained_on_dataset_v2_task_test_ds_rc.csv")

original_cols = ['model_key', 'model', 'transfer_learning', 'imgsz', 'epochs', 'batch',
                   'dataset', 'optimizer', 'lr0', 'all_P', 'all_R', 'all_F1', 'all_mAP@.5',
                   'all_mAP@.5:.95', 'FPS'] 

new_cols = ['model_key', 'Model', 'TL', 'imgsz', 'epochs', 'batch',
                   'dataset', 'optimizer', 'lr', 'Precision', 'Recall', 'all_F1', 'mAP@.5',
                   'mAP@.5:.95', 'FPS'] 

df_rc.rename(columns=dict(zip(original_cols, new_cols)), inplace=True)
df_rc["model_key"] = df_rc["model_key"].str.strip("3").str.strip("2")

In [5]:
df1 = df1.merge(df_rc[["model_key","FPS"]], on='model_key', how='left')

In [6]:
df3 = pd.concat([df1, df2])

cols = ['TL', 'imgsz', 'epochs', 'Precision', 'Recall', 'mAP@.5','mAP@.5:.95', 'FPS', 'optimization_technique'] 

# Convert a column to int
df3['imgsz'] = df3['imgsz'].astype(int)
df3['epochs'] = df3['epochs'].astype(int)

df3[cols].sort_values(by=['TL', 'imgsz', 'epochs', 'optimization_technique'], ascending=False)

TL  imgsz  epochs  Precision  Recall  mAP@.5  mAP@.5:.95       FPS  \
1  yolov8s    800     100      0.924   0.868   0.926       0.691  8.350000   
0  yolov8s    800     100      0.927   0.867   0.925       0.702  4.850000   
3  yolov8s    800     100      0.922   0.883   0.938       0.728  3.167564   
9  yolov8s    800      60      0.927   0.873   0.929       0.690  8.180000   
8  yolov8s    800      60      0.922   0.873   0.931       0.704  4.760000   
4  yolov8s    800      60      0.931   0.879   0.940       0.723  3.249919   
3  yolov8m    640     100      0.939   0.871   0.932       0.713  3.180000   
2  yolov8m    640     100      0.941   0.873   0.934       0.728  1.790000   
0  yolov8m    640     100      0.926   0.891   0.941       0.743  2.151000   
7  yolov8m    640      50      0.939   0.854   0.933       0.710  3.140000   
6  yolov8m    640      50      0.936   0.860   0.932       0.721  1.760000   
1  yolov8m    640      50      0.925   0.874   0.938       0.731  2.131742   
5  yolov8l    640      50      0.920   0.883   0.937       0.712  1.540000   
4  yolov8l    640      50      0.921   0.889   0.941       0.723  0.800000   
2  yolov8l    640      50      0.927   0.880   0.944       0.733  1.145475   

  optimization_technique  
1              Quantized  
0         Floating-point  
3                    NaN  
9              Quantized  
8         Floating-point  
4                    NaN  
3              Quantized  
2         Floating-point  
0                    NaN  
7              Quantized  
6         Floating-point  
1                    NaN  
5              Quantized  
4         Floating-point  
2                    NaN


The model with the highest performance in terms of Frames Per Second (FPS) is the 'yolov8s' with Quantized optimization technique and 800 imgsz (8.35 FPS).
The lowest performance is seen in the 'yolov8l' model with Floating-point optimization and 800 imgsz (0.8 FPS).


Both Floating-point and Quantized optimization techniques are used.
Quantized models tend to have higher FPS compared to Floating-point models.

The 'yolov8m' model without optimization, trained on 100 epochs and 640 imgsz has the highest recall (0.891).
The 'yolov8m' model with Floating-point optimization, trained on 100 epochs and 640 imgsz has the highest precision (0.941).

# Validation on randomized clips

In [9]:
# Load data from the JSON file
with open('optimization_results_on_rc.json', 'r') as json_file:
    data = json.load(json_file)


original_cols = ['Model', 'TL', 'imgsz', 'epochs', 'batch', 'dataset', 'optimizer', 'lr',
                   'optimization_technique', 'model_performance',
                   'stats.metrics/precision(B)', 'stats.metrics/recall(B)',
                   'stats.metrics/mAP50(B)', 'stats.metrics/mAP50-95(B)', 'stats.fitness']

new_cols = ['Model', 'TL', 'imgsz', 'epochs', 'batch', 'dataset', 'optimizer', 'lr',
                   'optimization_technique', 'FPS',
                   'Precision', 'Recall',
                   'mAP@.5', 'mAP@.5:.95', 'stats.fitness']

df2_rc = pd.json_normalize(data)
df2_rc.rename(columns=dict(zip(original_cols, new_cols)), inplace=True)

In [10]:
cols = ['TL', 'imgsz', 'epochs', 'Precision', 'Recall', 'mAP@.5','mAP@.5:.95', 'FPS', 'optimization_technique'] 

df3 = pd.concat([df_rc, df2_rc])

# Convert a column to int
df3['imgsz'] = df3['imgsz'].astype(int)
df3['epochs'] = df3['epochs'].astype(int)

df3[cols].sort_values(by=['TL', 'imgsz', 'epochs', 'optimization_technique'], ascending=False)

TL  imgsz  epochs  Precision    Recall    mAP@.5  mAP@.5:.95  \
1  yolov8s    800     100   0.274490  0.272843  0.180930    0.071675   
0  yolov8s    800     100   0.293009  0.282995  0.200955    0.080638   
3  yolov8s    800     100   0.578000  0.302000  0.314000    0.124000   
9  yolov8s    800      60   0.310603  0.227157  0.182933    0.074428   
8  yolov8s    800      60   0.307636  0.215736  0.196287    0.078507   
4  yolov8s    800      60   0.581000  0.293000  0.318000    0.125000   
3  yolov8m    640     100   0.422795  0.342640  0.281441    0.107850   
2  yolov8m    640     100   0.439115  0.337563  0.279439    0.109529   
1  yolov8m    640     100   0.576000  0.335000  0.366000    0.145000   
7  yolov8m    640      50   0.361666  0.324873  0.251282    0.099701   
6  yolov8m    640      50   0.344222  0.354061  0.251488    0.100194   
2  yolov8m    640      50   0.331000  0.345000  0.285000    0.114000   
5  yolov8l    640      50   0.570479  0.276422  0.318189    0.133311   
4  yolov8l    640      50   0.470349  0.326142  0.323435    0.135734   
0  yolov8l    640      50   0.481000  0.347000  0.309000    0.130000   

        FPS optimization_technique  
1  8.390000              Quantized  
0  4.920000         Floating-point  
3  3.167564                    NaN  
9  8.230000              Quantized  
8  4.810000         Floating-point  
4  3.249919                    NaN  
3  3.190000              Quantized  
2  1.800000         Floating-point  
1  2.151000                    NaN  
7  3.150000              Quantized  
6  1.780000         Floating-point  
2  2.131742                    NaN  
5  1.550000              Quantized  
4  0.880000         Floating-point  
0  1.145475                    NaN

In [11]:
df3[cols].sort_values(by=['Precision'])

TL  imgsz  epochs  Precision    Recall    mAP@.5  mAP@.5:.95  \
1  yolov8s    800     100   0.274490  0.272843  0.180930    0.071675   
0  yolov8s    800     100   0.293009  0.282995  0.200955    0.080638   
8  yolov8s    800      60   0.307636  0.215736  0.196287    0.078507   
9  yolov8s    800      60   0.310603  0.227157  0.182933    0.074428   
2  yolov8m    640      50   0.331000  0.345000  0.285000    0.114000   
6  yolov8m    640      50   0.344222  0.354061  0.251488    0.100194   
7  yolov8m    640      50   0.361666  0.324873  0.251282    0.099701   
3  yolov8m    640     100   0.422795  0.342640  0.281441    0.107850   
2  yolov8m    640     100   0.439115  0.337563  0.279439    0.109529   
4  yolov8l    640      50   0.470349  0.326142  0.323435    0.135734   
0  yolov8l    640      50   0.481000  0.347000  0.309000    0.130000   
5  yolov8l    640      50   0.570479  0.276422  0.318189    0.133311   
1  yolov8m    640     100   0.576000  0.335000  0.366000    0.145000   
3  yolov8s    800     100   0.578000  0.302000  0.314000    0.124000   
4  yolov8s    800      60   0.581000  0.293000  0.318000    0.125000   

        FPS optimization_technique  
1  8.390000              Quantized  
0  4.920000         Floating-point  
8  4.810000         Floating-point  
9  8.230000              Quantized  
2  2.131742                    NaN  
6  1.780000         Floating-point  
7  3.150000              Quantized  
3  3.190000              Quantized  
2  1.800000         Floating-point  
4  0.880000         Floating-point  
0  1.145475                    NaN  
5  1.550000              Quantized  
1  2.151000                    NaN  
3  3.167564                    NaN  
4  3.249919                    NaN

In [12]:
df3[cols].sort_values(by=['mAP@.5:.95'])

TL  imgsz  epochs  Precision    Recall    mAP@.5  mAP@.5:.95  \
1  yolov8s    800     100   0.274490  0.272843  0.180930    0.071675   
9  yolov8s    800      60   0.310603  0.227157  0.182933    0.074428   
8  yolov8s    800      60   0.307636  0.215736  0.196287    0.078507   
0  yolov8s    800     100   0.293009  0.282995  0.200955    0.080638   
7  yolov8m    640      50   0.361666  0.324873  0.251282    0.099701   
6  yolov8m    640      50   0.344222  0.354061  0.251488    0.100194   
3  yolov8m    640     100   0.422795  0.342640  0.281441    0.107850   
2  yolov8m    640     100   0.439115  0.337563  0.279439    0.109529   
2  yolov8m    640      50   0.331000  0.345000  0.285000    0.114000   
3  yolov8s    800     100   0.578000  0.302000  0.314000    0.124000   
4  yolov8s    800      60   0.581000  0.293000  0.318000    0.125000   
0  yolov8l    640      50   0.481000  0.347000  0.309000    0.130000   
5  yolov8l    640      50   0.570479  0.276422  0.318189    0.133311   
4  yolov8l    640      50   0.470349  0.326142  0.323435    0.135734   
1  yolov8m    640     100   0.576000  0.335000  0.366000    0.145000   

        FPS optimization_technique  
1  8.390000              Quantized  
9  8.230000              Quantized  
8  4.810000         Floating-point  
0  4.920000         Floating-point  
7  3.150000              Quantized  
6  1.780000         Floating-point  
3  3.190000              Quantized  
2  1.800000         Floating-point  
2  2.131742                    NaN  
3  3.167564                    NaN  
4  3.249919                    NaN  
0  1.145475                    NaN  
5  1.550000              Quantized  
4  0.880000         Floating-point  
1  2.151000                    NaN